### PR TITLE
Fix zero clipping in npred calculation

### PR DESCRIPTION
**Pull request by LauraOlivera**
This pull request addresses #3700 by moving the zero clipping applied to the predicted counts outside of the PSF convolution and into `dataset.npred`.

No tests are necessary because they were already added by #2342, which implemented the clipping in the first place:
https://github.com/gammapy/gammapy/blob/d342c5ac70cb321345f3c5e2c70f88ea8a116809/gammapy/datasets/tests/test_map.py#L250

With this fix, the snippet in #3700 results in

image

which is clipped where the negative values are. However, if the model amplitude is such that the negative model counts expected are less than the background counts (i.e. compatible with background fluctuations), then the result is this
image

which is not clipped to zero as it used to.

### CODE REVIEW

File: gammapy/datasets/map.py [hunk 1]
```diff
@@ -393,7 +393,7 @@ def npred(self):
 
         if self.background:
             npred_total += self.npred_background()
-
+        npred_total.data[npred_total.data < 0.0] = 0
```
Comment (registerrier) [inline gammapy/datasets/map.py:396]: Is the same command fine? Does `np.clip` provide better performance?

**Review by registerrier — COMMENTED** [review]
Thanks @LauraOlivera .

I have few comments:
- This will not fully solve the noise introduced by individual `fit_concolve`. Is it better to apply the filtering on the total npred compared to the npred per component?
- It might require a test since your specific problem went unnoticed: a negative flux should yield a negative npred at the `MapEvaluator` level no?

File: gammapy/datasets/map.py (hunk 1 above)
Comment (adonath) [reply gammapy/datasets/map.py:396]: I think @luca-giunti checked the implementation at the time and the "normal" masking was faster, if I remember correctly.

**Comment by adonath** [comment]
> This will not fully solve the noise introduced by individual fit_concolve. Is it better to apply the filtering on the total npred compared to the npred per component?

I think there is no solution to this, right? To be able to correctly computed "negative" npred for source components the npred cannot be clipped. Statistically the clipping on the total npred is correct, the positive npred is only required after the background is added.

**Comment by LauraOlivera** [comment]
Thanks @registerrier !

> It might require a test since your specific problem went unnoticed: a negative flux should yield a negative npred at the MapEvaluator level no?

Well I am not sure if we want to predict negative counts, since that is not really physical. The issue I was having is that in the absence of source, a the correct best-fit normalization might be (small) and negative, indicating compatible with a background fluctuation. Which was not possible with the existing implementation. So what I needed was `npred_signal` to be allowed to be negative, but the total `npred` should still be >=0 due to the background. I was of the understanding we don't want to allow negative `npred`. But of course I can change that...

**Comment by registerrier** [comment]
> Thanks @registerrier !
> 
> > It might require a test since your specific problem went unnoticed: a negative flux should yield a negative npred at the MapEvaluator level no?
> 
> Well I am not sure if we want to predict negative counts, since that is not really physical. The issue I was having is that in the absence of source, a the correct best-fit normalization might be (small) and negative, indicating compatible with a background fluctuation. Which was not possible with the existing implementation. So what I needed was `npred_signal` to be allowed to be negative, but the total `npred` should still be >=0 due to the background. I was of the understanding we don't want to allow negative `npred`. But of course I can change that...
> 
> EDIT: yeah, pretty much what Axel said... Only saw his comment after.

OK. Indeed, the correct way to remove fit noise is probably to do it at the end by removing possible negative npred.

Regarding the test, I was thinking to `MapEvaluator.compute_npred()`. This one should return negative entries if fed with a negative flux.

**Review by registerrier — APPROVED** [review]
Thanks @LauraOlivera ! This looks good to me.

**Review by adonath — APPROVED** [review]
Thanks @LauraOlivera, no further comment from my side either..

**Pull request merged**